salt/grains/core.py

```diff
@@ -2536,6 +2536,10 @@ def _hw_data(osdata):
                         grains[key] = salt.utils.stringutils.to_unicode(ifile.read().strip(), errors='replace')
                         if key == 'uuid':
                             grains['uuid'] = grains['uuid'].lower()
+                except UnicodeDecodeError:
```

Comment:
just one more thing for added clarification as i was re-looking at this change. Can you elaborate what is in the file that is causing the `UnicodeDecodeError`?  I am asking because I would assume that `salt.utils.stringutils.to_unicode` should be handling this properly? Just want to make sure we aren't just hiding an actual problem.

Reply:
Right. Can be that the catch is not in the correct place, I am not sure.

The problem is that there are some devices where the /sys/class/dmi/id/product_name file is not correct, is full of 0xFF content and some other random bytes. We reported this error as we are not sure if the problem is in the manufacturer or in the kernel.

In Python 3 installations this produces an UnicodeDecodeError exception in this point when the grain is try to read the file, and this makes the minion fails to connect to the master.

Maybe you can replicate this changing the code in the minion, and reading from a different file that contains non-valid unicode points

Reply:
Can you add a unit test similar to https://github.com/saltstack/salt/pull/53326/files#diff-3b5d15bc59b82fc8d4b15f819babf4faR1327 that reads in this type of data and ensure it raises the unicode error?

Reply:
Actually this is how I try to test first. Injecting invalid unicode data via `read_data`, but as this is mocking `salt.utils.files.fopen` is skipping where the problem happens: in the fopen instance itself.

```
Process Process-1:
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.6/site-packages/salt/scripts.py", line 157, in minion_process
    minion.start()
  File "/usr/lib/python3.6/site-packages/salt/cli/daemons.py", line 346, in start
    self._real_start()
  File "/usr/lib/python3.6/site-packages/salt/cli/daemons.py", line 358, in _real_start
    self.minion.tune_in()
  File "/usr/lib/python3.6/site-packages/salt/minion.py", line 1060, in tune_in
    self._spawn_minions()
  File "/usr/lib/python3.6/site-packages/salt/minion.py", line 1004, in _spawn_minions
    jid_queue=self.jid_queue)
  File "/usr/lib/python3.6/site-packages/salt/minion.py", line 976, in _create_minion_object
    jid_queue=jid_queue)
  File "/usr/lib/python3.6/site-packages/salt/minion.py", line 1130, in __init__
    self.opts['grains'] = salt.loader.grains(opts)
  File "/usr/lib/python3.6/site-packages/salt/loader.py", line 772, in grains
    ret = funcs[key]()
  File "/usr/lib/python3.6/site-packages/salt/grains/core.py", line 2088, in os_data
    grains.update(_hw_data(grains))
  File "/usr/lib/python3.6/site-packages/salt/grains/core.py", line 2529, in _hw_data
    grains[key] = salt.utils.stringutils.to_unicode(ifile.read().strip(), errors='replace')
  File "/usr/lib64/python3.6/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
```

From the stack trace we can suspect that the problem is in `ifile.read().strip()`, so this is not the object that we need to mock.

Reply:
thanks for the clarifications. would just like to get one more review here and should be good to go.